framework/mcp/README.md

```diff
@@ -0,0 +1,136 @@
+integration between LLM applications and external data sources and tools.
+
+MCP clients call servers by either invoking them as a subprocess and
+communicate with stdio, or by using SSE, which implements TCP socket
+communication.
+
+## What's Inside
+
+The Python programs have been derived from the [Writing MCP Clients] example
+program.
+
+- `example_builtin.py`: Exercise the basic built-in Model Context Protocol server
+  [@modelcontextprotocol/server-postgres] that provides read-only access to
+  PostgreSQL databases per `query` tool. This server enables LLMs to inspect
+  database schemas and execute read-only queries. It is written in TypeScript,
+  to be invoked with [npx].
+
+- `example_dbhub.py`: Exercise communication using [DBHub], a universal database
+  gateway implementing the Model Context Protocol (MCP) server interface. This
+  gateway allows MCP-compatible clients to connect to and explore different databases.
+
+- `example_jdbc.py`: Exercise JDBC communication using the [Model Context Protocol
+  Server for JDBC] from the [quarkus-mcp-servers] package, providing a range
+  of tools. It is written in Java, to be invoked with [JBang].
+
+## Resources
+
+- Read a [brief introduction to MCP] by ByteByteGo.
+- Read the canonical [Introduction to MCP].
+- Read about the [MCP Python SDK].
+- Read about configuring those servers with other MCP clients.
+  - [Claude Desktop configuration]
+  - [oterm configuration]
+- [Connecting to an already running MCP server] seems to become possible
+  with [MCP SSE]. As of 2025, not all servers implement that just yet.
+- [Model Context Protocol (MCP) @ CrateDB]
+
+## Setup
+
+Start a CrateDB instance on your machine.
+```shell
+docker run -it --rm \
+  --publish=4200:4200 --publish=5432:5432 \
+  --env=CRATE_HEAP_SIZE=2g \
+  crate:latest -Cdiscovery.type=single-node
+```
+
+Install the [uv] package manager and launcher.
+```shell
+{brew,pip} install uv
+```
+
+Install the [JBang] package manager and launcher.
+```shell
+{asdf,choco,dnf,brew,scoop,sdk} install jbang
+```
+
+## Install
+
+Acquire sources, set up sandbox, and install packages.
+```bash
+git clone https://github.com/crate/cratedb-examples
+cd cratedb-examples/framework/mcp
+uv pip install -r requirements.txt -r requirements-test.txt
+```
+
+## Synopsis
+
+```shell
+uv run example_builtin.py
+```
+```shell
+uv run example_dbhub.py
+```
+```shell
+uv run example_jdbc.py
+```
+
+## Tests
+
+Run integration tests.
+```bash
+pytest
+```
+
+## Development
+
+`ctk tail` is a handy command to follow the progress of CrateDB's `sys.jobs_log`,
+which is applicable for all sorts of driver, adapter, and connector explorations.
+After providing authentication information, just use uv's `uvx` launcher to invoke
+CrateDB Toolkit's tail command without installation.
+```shell
+export CRATEDB_SQLALCHEMY_URL=crate://crate@localhost:4200/
+```
+```shell
+uvx --from=cratedb-toolkit ctk tail -n 3 --follow --format=log-pretty sys.jobs_log
+```
```

Comment:
Just sharing a bit of development notes, now also derived into [`tail -f sys.jobs_log`](https://community.cratedb.com/t/tail-f-sys-jobs-log/2004).